bus/src/main/java/org/hawkular/metrics/component/publish/InsertedDataSubscriber.java

```diff
@@ -131,5 +149,43 @@ void shutdown() {
         if (subscription != null) {
             subscription.unsubscribe();
         }
+        publishTask.cancel();
+        publishTimer.cancel();
+    }
+
+    private int getPublishPeriod() {
+        try {
+            return Integer.parseInt(publishPeriodProperty);
+        } catch (NumberFormatException e) {
+            log.warnf("Invalid publish period. Setting default value %s", METRICS_PUBLISH_PERIOD.defaultValue());
+            return Integer.parseInt(METRICS_PUBLISH_PERIOD.defaultValue());
+        }
+    }
+
+    private class PublishTask extends TimerTask {
+        @Override
+        public void run() {
+            if (!pendingMetrics.isEmpty()) {
+                getAndClearPendingMetrics().stream().forEach(metric -> {
+                    if (metric.getMetricId().getType() == STRING) {
+                        log.warn("STRING type metrics are not yet supported for bus integration");
+                    } else if (metric.getMetricId().getType() == AVAILABILITY) {
+                        @SuppressWarnings("unchecked")
+                        Metric<AvailabilityType> avail = (Metric<AvailabilityType>) metric;
+                        publishAvailablility(avail);
+                    } else {
+                        @SuppressWarnings("unchecked")
+                        Metric<? extends Number> numeric = (Metric<? extends Number>) metric;
+                        publishNumeric(numeric);
+                    }
+                });
+            }
+        }
+    }
+
+    private synchronized List<Metric> getAndClearPendingMetrics() {
+        ArrayList<Metric> result = new ArrayList<>(pendingMetrics);
+        pendingMetrics.clear();
```

Comment:
This can cause a data loss. What's the point of synchronized method when pendingMetrics is not protected against concurrent modifications?

You should use a lock here that's the same as with the "onInsertedData" to prevent data being added between fetching the results and clearing.

Just ditch the synchronized keyword. It doesn't protect the pendingMetrics list here at all.